tests/unit/test_bake_project.py

```diff
@@ -1,30 +1,102 @@
 
 
-def test_run_flake8(cookies: t.Any) -> None:
-    result = cookies.bake(extra_context={"project_slug": "flake8-compat"})
-    with inside_dir(str(result.project)):
+def test_run_flake8(cookies: Cookies) -> None:
+    result = cookies.bake(extra_context={"project_dir": "flake8-compat"})
+    with inside_dir(str(result.project_path)):
         subprocess.check_call(["flake8"])
 
 
-def test_project_slug_regex_hook(cookies: t.Any) -> None:
-    result = cookies.bake(extra_context={"project_slug": "test_project"})
+def test_project_dir_hook(cookies: Cookies) -> None:
+    result = cookies.bake(extra_context={"project_dir": "myproject"})
+    assert result.exit_code == 0
+    result = cookies.bake(extra_context={"project_dir": "my-project"})
+    assert result.exit_code == 0
+    result = cookies.bake(extra_context={"project_dir": "my?project"})
+    assert result.exit_code != 0
+    if sys.platform == "win32":
+        # Unfortunately, pre_gen hook is called before cookiecutter copies the template
+        #  into the TMP dir for rendering.
+        # This will not hurt the user,
+        #  but the error message will also include a traceback
+        assert isinstance(result.exception, OSError)
+    else:
+        assert isinstance(result.exception, FailedHookException)
+    result = cookies.bake(extra_context={"project_dir": "t" * 256})
     assert result.exit_code != 0
-    assert isinstance(result.exception, FailedHookException)
 
 
-def test_project_slug_length_hook(cookies: t.Any) -> None:
-    result = cookies.bake(extra_context={"project_slug": "t" * 29})
-    assert result.exit_code != 0
-    assert isinstance(result.exception, FailedHookException)
+def test_project_id_hook(cookies: Cookies) -> None:
+    wrong_ids = [
+        "qwe/qwe",
+        "qwe?qwe",
+        "qwe!qwe",
+        "qwe.qwe",
+        "qwe%qwe",
+        "qwe-qwe",
+        "-qwe",
+        "qwe-",
+        "qwe-qwe",
+        "123",
+        "1 23",
+        "1qwe23",
+    ]
+    correct_ids = [
+        "qwe",
+        "q",
+        "qwe_qwe",
+        "_qwe",
+        "qwe_",
+        "qwe123",
+        "qwe_123",
+        "qwe" * 20,
+    ]
+    for id_ in wrong_ids:
+        result = cookies.bake(extra_context={"project_id": id_})
+        assert result.exit_code != 0, id_
+        assert isinstance(result.exception, FailedHookException)
+    for id_ in correct_ids:
+        result = cookies.bake(extra_context={"project_id": id_})
+        assert result.exit_code == 0, id_
+
+
+@pytest.mark.parametrize("preserve_comments", ["yes", "no"])
+def test_project_descr_with_comments(cookies: Cookies, preserve_comments: str) -> None:
+    result = cookies.bake(
+        extra_context={
+            "project_dir": "project-with-comments",
+            "preserve Neuro Flow template hints": preserve_comments,
+        }
+    )
+    assert result.exit_code == 0
+    l_com = "https://neu-ro.gitbook.io/neuro-flow/reference/live-workflow-syntax"
```

Comment:
the string to check seems fragile (may change in the future and break otherwise valid tests). can we check for presence lines with comment sign instead?

Reply:
good idea, made in c7d7b05